server/src/main/java/org/opensearch/index/engine/InternalEngine.java

```diff
@@ -2156,15 +2148,26 @@ protected SegmentInfos getLatestSegmentInfos() {
      */
     @Override
     public GatedCloseable<SegmentInfos> getSegmentInfosSnapshot() {
-        final SegmentInfos segmentInfos = getLatestSegmentInfos();
+        final SegmentInfos segmentInfos;
+        final OpenSearchDirectoryReader reader;
+        final GatedCloseable<IndexCommit> lastIndexCommit = acquireLastIndexCommit(false);
         try {
-            indexWriter.incRefDeleter(segmentInfos);
+            reader = internalReaderManager.acquire();
+            SegmentInfos readerInfos = ((StandardDirectoryReader) reader.getDelegate()).getSegmentInfos();
+            // if the latest index commit on disk has a higher gen than the reader load that instead.
+            // This ensures our commit file will exist throughout upload process.
+            if (lastIndexCommit.get().getGeneration() > readerInfos.getGeneration()) {
+                segmentInfos = SegmentInfos.readCommit(store.directory(), lastIndexCommit.get().getSegmentsFileName());
+            } else {
+                segmentInfos = readerInfos;
+            }
         } catch (IOException e) {
             throw new EngineException(shardId, e.getMessage(), e);
         }
         return new GatedCloseable<>(segmentInfos, () -> {
             try {
-                indexWriter.decRefDeleter(segmentInfos);
+                internalReaderManager.release(reader);
```

Comment:
If we are relying on last index commit to provide the segment infos (when if condition holds true), should we call the `internalReaderManager.release(reader)` earlier itself? Or are there any implications of releasing vs not releasing?

Reply:
good callout as well, we don't need the reader ref as the segments will be preserved by the commit.